src/devices/device.ts

```diff
@@ -78,16 +93,73 @@ export class Device extends EventEmitter {
         this.hub.send(data, characteristic, callback);
     }
 
-    public subscribe (mode: number) {
+    public subscribe (modeName: string) {
         this._ensureConnected();
-        if (mode !== this._mode) {
-            this._mode = mode;
-            this.hub.subscribe(this.portId, this.type, mode);
+        if (modeName !== this._mode) {
+            this._mode = modeName;
+
+            const modeNum = this._modes[modeName].num[this.hub.type];
+            if (modeNum === undefined) {
+                // TODO : error handling -> unsupported mode
+                return;
+            }
+            this.hub.subscribe(this.portId, this.type, modeNum);
         }
     }
 
     public receive (message: Buffer) {
-        this.emit("receive", message);
+        if (this._mode === undefined) {
+            // TODO : error handling -> no mode defined
+            return;
+        }
+
+        const mode = this._modes[this._mode];
+
+        if (!mode.values) {
+            // TODO : error handling -> no parsing informations
+            return;
+        }
+
+        const data = [];
+
+        for (let index = 0; index <= message.length; index += Consts.ValueBits[mode.values.type]) {
+            switch (mode.values.type) {
+                case Consts.ValueType.UInt8: {
+                    data.push(message.readUInt8(index));
+                    break;
+                }
+                case Consts.ValueType.Int8: {
+                    data.push(message.readInt8(index));
+                    break;
+                }
+                case Consts.ValueType.UInt16: {
+                    data.push(message.readUInt16LE(index));
+                    break;
+                }
+                case Consts.ValueType.Int16: {
+                    data.push(message.readInt16LE(index));
+                    break;
+                }
+                case Consts.ValueType.UInt32: {
+                    data.push(message.readUInt32LE(index));
+                    break;
+                }
+                case Consts.ValueType.Int32: {
+                    data.push(message.readInt32LE(index));
+                    break;
+                }
+                case Consts.ValueType.Float: {
+                    data.push(message.readFloatLE(index));
+                    break;
+                }
+            }
+        }
+
+        if (mode.event) {
+            this.emit(mode.event, ...data);
+        }
```

Comment:
emit related event if any.

This break the distance one as it done not apply calculation on the value and also the colorAndDistance which will receive color, distance (0-10), led color and reflectivity instead of only color and distance.